modules/gapi/src/executor/gthreadedexecutor.cpp

```diff
@@ -0,0 +1,525 @@
+        {
+            // Let island allocate it's outputs if it can,
+            // allocate cv::Mat and wrap it with RMat otherwise
+            GAPI_Assert(!nh->inNodes().empty());
+            const auto desc = util::get<cv::GMatDesc>(d.meta);
+            auto& exec = m_gim.metadata(nh->inNodes().front()).get<IslandExec>().object;
+            auto& rmat = m_res.slot<cv::RMat>()[d.rc];
+            if (exec->allocatesOutputs()) {
+                rmat = exec->allocate(desc);
+            } else {
+                Mat mat;
+                createMat(desc, mat);
+                rmat = make_rmat<RMatOnMat>(mat);
+            }
+        }
+        break;
+
+    case GShape::GSCALAR:
+        if (d.storage == Data::Storage::CONST_VAL)
+        {
+            auto rc = RcDesc{d.rc, d.shape, d.ctor};
+            magazine::bindInArg(m_res, rc, m_gm.metadata(orig_nh).get<ConstValue>().arg);
+        }
+        break;
+
+    case GShape::GARRAY:
+        if (d.storage == Data::Storage::CONST_VAL)
+        {
+            auto rc = RcDesc{d.rc, d.shape, d.ctor};
+            magazine::bindInArg(m_res, rc, m_gm.metadata(orig_nh).get<ConstValue>().arg);
+        }
+        break;
+    case GShape::GOPAQUE:
+        // Constructed on Reset, do nothing here
+        break;
+    case GShape::GFRAME: {
+        // Should be defined by backend, do nothing here
+        break;
+    }
+    default:
+        GAPI_Error("InternalError");
+    }
+}
+
+cv::gimpl::IslandActor::IslandActor(const std::vector<RcDesc>          &in_objects,
+                                    const std::vector<RcDesc>          &out_objects,
+                                    std::shared_ptr<GIslandExecutable> isl_exec,
+                                    Mag                                &res,
+                                    std::mutex                         &m)
+    : m_in_objs(in_objects),
+      m_out_objs(out_objects),
+      m_isl_exec(isl_exec),
+      m_res(res),
+      m_mutex(m) {
+}
+
+void cv::gimpl::IslandActor::run() {
+    cv::gimpl::GThreadedExecutor::Input  i{m_res,  m_in_objs,  m_mutex};
+    cv::gimpl::GThreadedExecutor::Output o{m_res,  m_out_objs, m_mutex};
+    m_isl_exec->run(i, o);
+    // NB: Can't throw an exception there as it's supposed to be
+    // executed in different threads
+    try {
+        o.verify();
+    } catch (...) {
+        m_e = std::current_exception();
+    }
+}
+
+void cv::gimpl::IslandActor::verify() {
+    if (m_e) {
+        std::rethrow_exception(m_e);
+    }
+};
+
+class cv::gimpl::Task {
+public:
+    using F = std::function<void()>;
+    using Ptr = std::shared_ptr<Task>;
+
+    Task(F f, std::vector<Task::Ptr> &&deps);
+
+    void run();
+    void reset() { m_ready_deps.store(0u); }
+
+private:
+    F                     m_f;
+    const uint32_t        m_num_deps;
+    std::atomic<uint32_t> m_ready_deps;
+    std::vector<Task*>    m_dependents;
+};
+
+cv::gimpl::Task::Task(F f, std::vector<Task::Ptr> &&deps)
```

Comment:
..here rename `deps` to `producers`?

Reply:
Please, correct me, but how `deps` here are `producers`? IMHO they are still the same consumers of the current task

Reply:
From my point of view, the `dependencies` is better name since it describes relationship between tasks.
The one task may depend on several other tasks but task itself doesn't consume or produce other tasks.